packages/block-editor/src/components/block-list-appender/style.scss

```diff
@@ -10,26 +10,32 @@
 		max-width: none;
 	}
 
-	// Add a little left margin when used horizontally.
-	// The right margin should be set to auto, so as to not shift layout in flex containers.
-	margin: 0 auto 0 $grid-unit-10;
-
-	// ... unless it's the only child.
-	&:first-child {
-		margin-left: 0;
-	}
-
 	.block-editor-default-block-appender {
 		margin: $grid-unit-10 0;
 	}
 
-	// Animate appearance.
+	// Add an explicit left margin of zero and auto right margin to work in horizontal
+	// flex containers. Without it, a "space-between"-like effect from two auto margins
+	// will cause the black plus to sit in the center of what space is left.
+	margin: 0 auto 0 0;
+
+	// Black square plus appender.
 	.block-list-appender__toggle {
 		padding: 0;
+
+		// Animate appearance.
 		opacity: 1;
 		transform: scale(1);
 		transition: all 0.1s ease;
 		@include reduce-motion("transition");
+
+		// The black square button should have a little left margin in horizontal containers.
+		margin-left: $grid-unit-10;
+	}
+
+	// Cancel any left margin if the black plus is the only child.
+	&:first-of-type .block-list-appender__toggle {
```

Comment:
I don't really understand this part, I might be missing some context. Why will it be the only child if it matches `:first-of-type`?

Reply:
Excellent question, thanks for calling it out.

`first-of-type` is used because the appender is not the `first-child` in an otherwise "empty" navigation block:

<img width="1811" alt="Screenshot 2021-07-01 at 11 14 48" src="https://user-images.githubusercontent.com/1204802/124099632-06c09f00-da5e-11eb-9278-9c9c7f9b30a0.png">

Shown here with zero width is an actual placeholder state for the navigation block. That placeholder state exists here but has zero with so that it can ensure the height of the navigation block is the same between the empty state, and when there are menu items inside, ensuring no layout shift when you insert.

Observe here in the top right corner that the `margin-left: 0;` rule is correctly applied to the appender when the `first-of-type` rule hits the target:

<img width="1811" alt="Screenshot 2021-07-01 at 11 17 31" src="https://user-images.githubusercontent.com/1204802/124100022-60c16480-da5e-11eb-866c-8004fe2fb92c.png">

Here, it's no longer the `first-of-type`, after a menu item has been inserted, and the 8px left margin applies:

<img width="1811" alt="Screenshot 2021-07-01 at 11 21 13" src="https://user-images.githubusercontent.com/1204802/124100244-98c8a780-da5e-11eb-8d0c-379d51cc51ec.png">

That same rule applies here as well:

<img width="1811" alt="Screenshot 2021-07-01 at 11 22 41" src="https://user-images.githubusercontent.com/1204802/124100327-ada53b00-da5e-11eb-8686-6d3fb8366e01.png">

As noted in https://github.com/WordPress/gutenberg/pull/33087#issuecomment-871227848, I think ultimately we need to remove the black plus from the editing canvas entirely, make it a popover that exists outside the canvas/iframe and is positioned in a way that doesn't cause layout shifts. That will be an opportunity to tear out all these complex rules.

In the mean time, I will add a comment to elaborate on why first-of-type is used. Sound good?